File: headless-server/youtube_service.py
import os
from googleapiclient.http import MediaFileUpload

class YouTubeService:
    """YouTube service for headless server operation"""
    
    def __init__(self, youtube_client, auth=None):
        self.youtube = youtube_client
        self.auth = auth
    
    def create_broadcast(self, broadcast_data):
        """Create YouTube live broadcast"""
        try:
            body = {
                "snippet": {
                    "title": broadcast_data["title"],
                    "description": broadcast_data["description"],
                    "scheduledStartTime": broadcast_data["scheduledStartTime"],
                    "tags": broadcast_data["tags"],
                    "categoryId": broadcast_data["categoryId"]
                },
                "status": {
                    "privacyStatus": broadcast_data["privacyStatus"],
                    "selfDeclaredMadeForKids": broadcast_data.get("madeForKids", False)
                },
                "contentDetails": {
                    "enableDvr": broadcast_data.get("enableDvr", True),
                    "enableEmbed": broadcast_data.get("enableEmbed", True),
                    "recordFromStart": broadcast_data.get("recordFromStart", True),
                    "enableAutoStart": True,
                    "enableAutoStop": True
                }
            }
            
            if self.auth and self.auth.selected_channel_id:
                body["snippet"]["channelId"] = self.auth.selected_channel_id
            
            request = self.youtube.liveBroadcasts().insert(
                part="snippet,status,contentDetails",
                body=body
            )
            response = request.execute()
            
            broadcast_id = response["id"]
            
            if broadcast_data.get("containsSyntheticMedia", False):
                self.update_video_synthetic_media(broadcast_id, True)
            
            if broadcast_data.get("enableMonetization", False):
                self.update_video_monetization(broadcast_id, True)
            
            return True, response
        except Exception as e:
            return False, str(e)
    
    def update_video_synthetic_media(self, video_id, contains_synthetic):
        """Update synthetic media flag"""
        try:
            request = self.youtube.videos().update(
                part="status",
                body={
                    "id": video_id,
                    "status": {
                        "containsSyntheticMedia": contains_synthetic
                    }
                }
            )
            response = request.execute()
            return True, response
        except Exception as e:
            return False, str(e)
    
    def update_video_monetization(self, video_id, enable_monetization):
        """Set video as eligible for monetization"""
        try:
            get_request = self.youtube.videos().list(
                part="status",
                id=video_id
            )
            get_response = get_request.execute()
            
            if not get_response.get("items"):
                return False, f"Video ID not found: {video_id}"
            
            current_status = get_response["items"][0]["status"]
            
            update_body = {
                "id": video_id,
                "status": {
                    "privacyStatus": current_status.get("privacyStatus", "public"),
                    "madeForKids": False,
                    "selfDeclaredMadeForKids": False,
                }
            }
            
            if "embeddable" in current_status:
                update_body["status"]["embeddable"] = current_status["embeddable"]
            if "license" in current_status:
                update_body["status"]["license"] = current_status["license"]
            if "publicStatsViewable" in current_status:
                update_body["status"]["publicStatsViewable"] = current_status["publicStatsViewable"]
            
            request = self.youtube.videos().update(
                part="status",
                body=update_body
            )
            response = request.execute()
            return True, response
        except Exception as e:
            return False, str(e)
    
    def create_stream(self, title, latency="normal"):
        """Create live stream"""
        try:
            stream_title = f"Stream: {title}"
            
            request = self.youtube.liveStreams().insert(
                part="snippet,cdn,contentDetails",
                body={
                    "snippet": {
                        "title": stream_title
                    },
                    "cdn": {
                        "frameRate": "variable",
                        "ingestionType": "rtmp",
                        "resolution": "variable"
                    },
                    "contentDetails": {
                        "isReusable": False
                    }
                }
            )
            response = request.execute()
            return True, response
        except Exception as e:
            return False, str(e)
    
    def get_stream_info(self, stream_id):
        """Get stream information"""
        try:
            request = self.youtube.liveStreams().list(
                part="snippet,cdn,status",
                id=stream_id
            )
            response = request.execute()
            if response.get("items"):
                return True, response["items"][0]
            else:
                return False, f"Stream ID not found: {stream_id}"
        except Exception as e:
            return False, str(e)
    
    def find_stream_by_key(self, stream_key):
        """Find stream by key"""
        try:
            request = self.youtube.liveStreams().list(
                part="snippet,cdn,status",
                mine=True,
                maxResults=50
            )
            response = request.execute()
            
            for stream in response.get("items", []):
                if stream.get("cdn", {}).get("ingestionInfo", {}).get("streamName") == stream_key:
                    return True, stream
            
            return False, f"No stream found with key: {stream_key}"
        except Exception as e:
            return False, str(e)
    
    def bind_broadcast_to_stream(self, broadcast_id, stream_id):
        """Bind broadcast to stream"""
        try:
            request = self.youtube.liveBroadcasts().bind(
                part="id,contentDetails",
                id=broadcast_id,
                streamId=stream_id
            )
            response = request.execute()
            return True, response
        except Exception as e:
            return False, str(e)
    
    def upload_thumbnail(self, broadcast_id, thumbnail_path):
        """Upload thumbnail"""
        if not os.path.exists(thumbnail_path):
            return False, f"Thumbnail file not found: {thumbnail_path}"
        
        try:
            request = self.youtube.thumbnails().set(
                videoId=broadcast_id,
                media_body=MediaFileUpload(thumbnail_path, chunksize=-1, resumable=True)
            )
            response = request.execute()
            return True, response
        except Exception as e:
            return False, str(e)
    
    def process_broadcast(self, broadcast_data, log_callback=None):
        """Process complete broadcast creation"""
        def log(msg):
            if log_callback:
                log_callback(msg)
            print(msg)
        
        title = broadcast_data["title"]
        log(f"[BROADCAST] Processing: {title}")
        
        try:
            success, result = self.create_broadcast(broadcast_data)
            if not success:
                log(f"[BROADCAST] ✗ Failed to create: {result}")
                return False, result
            
            broadcast_id = result["id"]
            log(f"[BROADCAST] ✓ Created: {broadcast_id}")
            
            if broadcast_data.get("containsSyntheticMedia", False):
                log(f"[BROADCAST] Setting synthetic media flag...")
            
            if broadcast_data.get("enableMonetization", False):
                log(f"[BROADCAST] Setting monetization eligibility...")
            
            # Handle stream
            if broadcast_data.get("streamId"):
                stream_id = broadcast_data["streamId"]
                log(f"[BROADCAST] Using existing stream: {stream_id}")
                
                success, result = self.get_stream_info(stream_id)
                if not success:
                    log(f"[BROADCAST] ✗ Stream verification failed: {result}")
                    return False, result
                log(f"[BROADCAST] ✓ Stream verified")
            elif broadcast_data.get("streamKey"):
                stream_key = broadcast_data["streamKey"]
                log(f"[BROADCAST] Looking up stream key: {stream_key}")
                
                success, result = self.find_stream_by_key(stream_key)
                if not success:
                    log(f"[BROADCAST] ✗ Stream lookup failed: {result}")
                    return False, result
                
                stream_id = result["id"]
                log(f"[BROADCAST] ✓ Found stream: {stream_id}")
            else:
                success, result = self.create_stream(
                    title, 
                    broadcast_data.get("latency", "normal")
                )
                if not success:
                    log(f"[BROADCAST] ✗ Failed to create stream: {result}")
                    return False, result
                
                stream_id = result["id"]
                log(f"[BROADCAST] ✓ Created stream: {stream_id}")
            
            success, result = self.bind_broadcast_to_stream(broadcast_id, stream_id)
            if not success:
                log(f"[BROADCAST] ✗ Failed to bind: {result}")
                return False, result
            
            log(f"[BROADCAST] ✓ Bound to stream")
            
            thumbnail_path = broadcast_data.get("thumbnailPath")
            if thumbnail_path and thumbnail_path.strip():
                success, result = self.upload_thumbnail(broadcast_id, thumbnail_path)
                if success:
                    log(f"[BROADCAST] ✓ Uploaded thumbnail")
                else:
                    log(f"[BROADCAST] ⚠ Thumbnail upload failed: {result}")
            
            log(f"[BROADCAST] ✓ Completed: {title}\n")
            return True, broadcast_id
        except Exception as e:
            log(f"[BROADCAST] ✗ Unexpected error: {str(e)}")
            return False, str(e)
    
    def get_upcoming_broadcasts(self):
        """Get upcoming broadcasts"""
        try:
            request = self.youtube.liveBroadcasts().list(
                part="snippet,status,contentDetails",
                broadcastStatus="upcoming",
                maxResults=50
            )
            response = request.execute()
            return True, response.get("items", [])
        except Exception as e:
            return False, str(e)
    
    def upload_video_file(self, video_data, progress_callback=None):
        """Upload video file"""
        try:
            video_path = video_data["videoPath"]
            
            body = {
                "snippet": {
                    "title": video_data["title"],
                    "description": video_data["description"],
                    "tags": video_data["tags"],
                    "categoryId": video_data["categoryId"]
                },
                "status": {
                    "privacyStatus": video_data["privacyStatus"],
                    "selfDeclaredMadeForKids": video_data.get("madeForKids", False)
                }
            }
            
            if video_data.get("publishAt"):
                body["status"]["publishAt"] = video_data["publishAt"]
            
            if self.auth and self.auth.selected_channel_id:
                body["snippet"]["channelId"] = self.auth.selected_channel_id
            
            media = MediaFileUpload(
                video_path,
                chunksize=1024*1024*5,
                resumable=True
            )
            
            request = self.youtube.videos().insert(
                part="snippet,status",
                body=body,
                media_body=media
            )
            
            response = None
            while response is None:
                status, response = request.next_chunk()
                if status:
                    progress = int(status.progress() * 100)
                    if progress_callback:
                        progress_callback(progress)
                    print(f"[UPLOAD] Progress: {progress}%")
            
            video_id = response["id"]
            
            if video_data.get("containsSyntheticMedia", False):
                self.update_video_synthetic_media(video_id, True)
            
            if video_data.get("enableMonetization", False):
                self.update_video_monetization(video_id, True)
            
            return True, video_id
        except Exception as e:
            return False, str(e)

File: webstreampro/src/services/YouTubeService.js
const fs = require('fs').promises;

class YouTubeService {
  constructor(authService) {
    this.authService = authService;
  }

  getYouTube() {
    return this.authService.getYouTubeClient();
  }

  async createBroadcast(broadcastData) {
    try {
      const youtube = this.getYouTube();
      
      const body = {
        snippet: {
          title: broadcastData.title,
          description: broadcastData.description,
          scheduledStartTime: broadcastData.scheduledStartTime,
          tags: broadcastData.tags || [],
          categoryId: broadcastData.categoryId || '20'
        },
        status: {
          privacyStatus: broadcastData.privacyStatus || 'public',
          selfDeclaredMadeForKids: broadcastData.madeForKids || false
        },
        contentDetails: {
          enableDvr: broadcastData.enableDvr !== false,
          enableEmbed: broadcastData.enableEmbed !== false,
          recordFromStart: broadcastData.recordFromStart !== false,
          enableAutoStart: true,
          enableAutoStop: true
        }
      };

      const selectedChannelId = this.authService.getSelectedChannelId();
      if (selectedChannelId) {
        body.snippet.channelId = selectedChannelId;
      }

      const response = await youtube.liveBroadcasts.insert({
        part: 'snippet,status,contentDetails',
        requestBody: body
      });

      const broadcastId = response.data.id;

      if (broadcastData.containsSyntheticMedia) {
        await this.updateVideoSyntheticMedia(broadcastId, true);
      }

      if (broadcastData.enableMonetization) {
        await this.updateVideoMonetization(broadcastId, true);
      }

      return { success: true, data: response.data };
    } catch (error) {
      console.error('Error creating broadcast:', error);
      return { success: false, error: error.message };
    }
  }

  async updateVideoSyntheticMedia(videoId, containsSynthetic) {
    try {
      const youtube = this.getYouTube();
      
      const response = await youtube.videos.update({
        part: 'status',
        requestBody: {
          id: videoId,
          status: {
            containsSyntheticMedia: containsSynthetic
          }
        }
      });

      return { success: true, data: response.data };
    } catch (error) {
      console.error('Error updating synthetic media:', error);
      return { success: false, error: error.message };
    }
  }

  async updateVideoMonetization(videoId, enableMonetization) {
    try {
      const youtube = this.getYouTube();
      
      const getResponse = await youtube.videos.list({
        part: 'status',
        id: videoId
      });

      if (!getResponse.data.items || getResponse.data.items.length === 0) {
        throw new Error(`Video ID not found: ${videoId}`);
      }

      const currentStatus = getResponse.data.items[0].status;

      const updateBody = {
        id: videoId,
        status: {
          privacyStatus: currentStatus.privacyStatus || 'public',
          madeForKids: false,
          selfDeclaredMadeForKids: false
        }
      };

      if (currentStatus.embeddable !== undefined) {
        updateBody.status.embeddable = currentStatus.embeddable;
      }
      if (currentStatus.license !== undefined) {
        updateBody.status.license = currentStatus.license;
      }
      if (currentStatus.publicStatsViewable !== undefined) {
        updateBody.status.publicStatsViewable = currentStatus.publicStatsViewable;
      }

      const response = await youtube.videos.update({
        part: 'status',
        requestBody: updateBody
      });

      return { success: true, data: response.data };
    } catch (error) {
      console.error('Error updating monetization:', error);
      return { success: false, error: error.message };
    }
  }

  async createStream(title, latency = 'normal') {
    try {
      const youtube = this.getYouTube();
      const streamTitle = `Stream: ${title}`;

      const response = await youtube.liveStreams.insert({
        part: 'snippet,cdn,contentDetails',
        requestBody: {
          snippet: {
            title: streamTitle
          },
          cdn: {
            frameRate: 'variable',
            ingestionType: 'rtmp',
            resolution: 'variable'
          },
          contentDetails: {
            isReusable: false
          }
        }
      });

      return { success: true, data: response.data };
    } catch (error) {
      console.error('Error creating stream:', error);
      return { success: false, error: error.message };
    }
  }

  async getStreamInfo(streamId) {
    try {
      const youtube = this.getYouTube();

      const response = await youtube.liveStreams.list({
        part: 'snippet,cdn,status',
        id: streamId
      });

      if (response.data.items && response.data.items.length > 0) {
        return { success: true, data: response.data.items[0] };
      } else {
        return { success: false, error: `Stream ID not found: ${streamId}` };
      }
    } catch (error) {
      console.error('Error getting stream info:', error);
      return { success: false, error: error.message };
    }
  }

  async findStreamByKey(streamKey) {
    try {
      const youtube = this.getYouTube();

      const response = await youtube.liveStreams.list({
        part: 'snippet,cdn,status',
        mine: true,
        maxResults: 50
      });

      for (const stream of response.data.items || []) {
        if (stream.cdn?.ingestionInfo?.streamName === streamKey) {
          return { success: true, data: stream };
        }
      }

      return { success: false, error: `No stream found with key: ${streamKey}` };
    } catch (error) {
      console.error('Error finding stream by key:', error);
      return { success: false, error: error.message };
    }
  }

  async bindBroadcastToStream(broadcastId, streamId) {
    try {
      const youtube = this.getYouTube();

      const response = await youtube.liveBroadcasts.bind({
        part: 'id,contentDetails',
        id: broadcastId,
        streamId: streamId
      });

      return { success: true, data: response.data };
    } catch (error) {
      console.error('Error binding broadcast to stream:', error);
      return { success: false, error: error.message };
    }
  }

  async uploadThumbnail(broadcastId, thumbnailPath) {
    try {
      const youtube = this.getYouTube();
      
      const fileContent = await fs.readFile(thumbnailPath);

      const response = await youtube.thumbnails.set({
        videoId: broadcastId,
        media: {
          body: fileContent
        }
      });

      return { success: true, data: response.data };
    } catch (error) {
      console.error('Error uploading thumbnail:', error);
      return { success: false, error: error.message };
    }
  }

  async processBroadcast(broadcastData, logCallback = null) {
    const log = (msg) => {
      console.log(msg);
      if (logCallback) logCallback(msg);
    };

    const title = broadcastData.title;
    log(`Processing: ${title}`);

    try {
      const createResult = await this.createBroadcast(broadcastData);
      if (!createResult.success) {
        log(`[X] Failed to create broadcast: ${createResult.error}`);
        return { success: false, error: createResult.error };
      }

      const broadcastId = createResult.data.id;
      log(`[OK] Created broadcast: ${broadcastId}`);

      if (broadcastData.containsSyntheticMedia) {
        log(`[INFO] Setting synthetic media flag...`);
      }

      if (broadcastData.enableMonetization) {
        log(`[INFO] Setting video as eligible for monetization...`);
      }

      let streamId;
      if (broadcastData.streamId) {
        streamId = broadcastData.streamId;
        log(`[INFO] Using existing stream ID: ${streamId}`);

        const streamInfo = await this.getStreamInfo(streamId);
        if (!streamInfo.success) {
          log(`[X] Stream ID verification failed: ${streamInfo.error}`);
          return { success: false, error: streamInfo.error };
        }
        log(`[OK] Stream verified: ${streamId}`);
      } else if (broadcastData.streamKey) {
        const streamKey = broadcastData.streamKey;
        log(`[INFO] Looking up stream with key: ${streamKey}`);

        const findResult = await this.findStreamByKey(streamKey);
        if (!findResult.success) {
          log(`[X] Stream lookup failed: ${findResult.error}`);
          return { success: false, error: findResult.error };
        }

        streamId = findResult.data.id;
        log(`[OK] Found stream: ${streamId}`);
      } else {
        const createStreamResult = await this.createStream(
          title,
          broadcastData.latency || 'normal'
        );
        if (!createStreamResult.success) {
          log(`[X] Failed to create stream: ${createStreamResult.error}`);
          return { success: false, error: createStreamResult.error };
        }

        streamId = createStreamResult.data.id;
        log(`[OK] Created stream: ${streamId}`);
      }

      const bindResult = await this.bindBroadcastToStream(broadcastId, streamId);
      if (!bindResult.success) {
        log(`[X] Failed to bind: ${bindResult.error}`);
        return { success: false, error: bindResult.error };
      }

      log(`[OK] Bound broadcast to stream`);

      if (broadcastData.thumbnailPath) {
        const uploadResult = await this.uploadThumbnail(broadcastId, broadcastData.thumbnailPath);
        if (uploadResult.success) {
          log(`[OK] Uploaded thumbnail`);
        } else {
          log(`[WARN] Thumbnail upload failed: ${uploadResult.error}`);
        }
      }

      log(`[OK] Completed: ${title}\n`);
      return { success: true, broadcastId };
    } catch (error) {
      log(`[X] Unexpected error: ${error.message}`);
      return { success: false, error: error.message };
    }
  }

  async getUpcomingBroadcasts() {
    try {
      const youtube = this.getYouTube();

      const response = await youtube.liveBroadcasts.list({
        part: 'snippet,status,contentDetails',
        broadcastStatus: 'upcoming',
        maxResults: 50
      });

      return { success: true, data: response.data.items || [] };
    } catch (error) {
      console.error('Error getting upcoming broadcasts:', error);
      return { success: false, error: error.message };
    }
  }

  async uploadVideoFile(videoData, onProgress = null) {
    try {
      const youtube = this.getYouTube();
      const videoPath = videoData.videoPath;

      const body = {
        snippet: {
          title: videoData.title,
          description: videoData.description,
          tags: videoData.tags || [],
          categoryId: videoData.categoryId || '20'
        },
        status: {
          privacyStatus: videoData.privacyStatus || 'public',
          selfDeclaredMadeForKids: videoData.madeForKids || false
        }
      };

      if (videoData.publishAt) {
        body.status.publishAt = videoData.publishAt;
      }

      const selectedChannelId = this.authService.getSelectedChannelId();
      if (selectedChannelId) {
        body.snippet.channelId = selectedChannelId;
      }

      const fileSize = (await fs.stat(videoPath)).size;
      const fileStream = require('fs').createReadStream(videoPath);

      const response = await youtube.videos.insert({
        part: 'snippet,status',
        requestBody: body,
        media: {
          body: fileStream
        }
      }, {
        onUploadProgress: (evt) => {
          const progress = (evt.bytesRead / fileSize) * 100;
          if (onProgress) {
            onProgress(Math.round(progress));
          }
        }
      });

      const videoId = response.data.id;

      if (videoData.containsSyntheticMedia) {
        await this.updateVideoSyntheticMedia(videoId, true);
      }

      if (videoData.enableMonetization) {
        await this.updateVideoMonetization(videoId, true);
      }

      return { success: true, videoId };
    } catch (error) {
      console.error('Error uploading video:', error);
      return { success: false, error: error.message };
    }
  }
}

module.exports = YouTubeService;
